### ISSUE TITLE
Haskell process command errored with: (end-of-buffer)

### ISSUE THREAD

**Issue by shanemikel** [haskell/haskell-mode#1260]
Hmm, every now and then I'm hitting this error.  I don't know why it comes up exactly, and peeking at the *haskell-process-log* the haskell process is still doing expected things AFAICT.  If anybody wants to help walk me through debugging this, let me know.

**Comment by shanemikel** [comment]
Actually, I can describe part of the situation I encounter it, if it helps.
So far, I've only noticed it in folding-mode, but sometimes the interactive mode and folding-mode seem to work together just fine.

On reload, I was getting the error, until I added an import that was missing to my source, and interactive mode started working again.

I've also checked this without folding-mode, and the error of missing module import is, as expected, printed to the ghci buffer.

I will try and keep from changing my emacs config too much, and I'll keep the haskell source around until somebody gets back to me.

**Comment by fice-t** [comment]
Can you get the error to trigger with `debug-on-error`?

**Comment by shanemikel** [comment]
You mean, isolate it with
(setq debug-on-error '(end-of-buffer))
?

**Comment by fice-t** [comment]
That or just setting it to `t`.

**Comment by shanemikel** [comment]
http://lpaste.net/664605885082894336

If you'd like, I can copy the output here

**Comment by shanemikel** [comment]
My debug-on-error is already set to t, I think the error is getting caught, and haskell-mode is printing that line to the message buffer

**Comment by shanemikel** [comment]
http://lpaste.net/336486210622980096

Okay, I think that's the info we want.

**Comment by fice-t** [comment]
That's odd, `condition-case-unless-debug` should trigger the debugger (that should be fixed if it's broken). Can you replace the following command in `haskell-process`, recompile and show the backtrace? That might be easier to see:
```elisp
(defun haskell-command-exec-complete (command response)
  "Call the command's complete function."
  (let ((comp-func (haskell-command-complete command)))
    (when comp-func
      (funcall comp-func
               (haskell-command-state command)
               response))))
```

**Comment by shanemikel** [comment]
Do you like the way I've been conducting the traces thus far?  If so, which function should I trace?

**Comment by fice-t** [comment]
If you replace that function then the debugger should just pop up on your error without an explicit trace. (Since that `condition-case` is the only place where that error message is located.)

**Comment by fice-t** [comment]
Hmm, I found an interesting variable `debug-ignored-errors` which includes `end-of-buffer`. Instead of doing the above (though that should work as well), try setting the that variable to `nil` and try again.

**Comment by shanemikel** [comment]
http://lpaste.net/4466009369221267456

**Comment by shanemikel** [comment]
I'm pinging this thread because I started that last comment before yours @fice-t.

**Comment by fice-t** [comment]
`haskell-check-paint-overlay` seems benign enough. Could you post `Ex.hs`?

**Comment by shanemikel** [comment]
Ex.hs
http://lpaste.net/8168584496237510656

OpLaws.hs
http://lpaste.net/826699333813403648

MyNonEmpty.hs
http://lpaste.net/7133813629228417024

**Comment by fice-t** [comment]
Odd: I can't reproduce the error with `haskell-check-paint-overlay` when using the same arguments.

The only function in there that would trigger that warning would be the `forward-char` bit.

**Comment by shanemikel** [comment]
I'm also using stack-ghci

**Comment by shanemikel** [comment]
yes I encountered the same error

**Comment by fice-t** [comment]
Could you try turning on debug/edebug for `haskell-check-paint-overlay` and stepping through the expression above? It should error at `(forward-char (1- coln))`, but I don't see why it would with those arguments...

**Comment by shanemikel** [comment]
Yeah, are you trying this with folding-mode? That's https://github.com/peterwang/folding.el or the `folding` package on melpa.

I am getting the error at `(forward-char (1- coln))`; I don't know how the other folding modes work, but this one actually changes the line numbers shown in the buffer.  If I debug this with all folds open (even with folding-mode enabled) I don't get this error; if any of them are closed, the line numbers are changed, and so the error happens.

So, I guess we need to decide if haskell-mode needs to work around this, or if folding-mode can do something to keep the line numbers intact.
In the meantime, is it possible for me to (cleanly) disable some behavior of haskell-mode and prevent this problem?

**Comment by shanemikel** [comment]
I'm sorry @fice-t, that I didn't make it clearer before you spent all this time with me.

**Comment by shanemikel** [comment]
Actually, do you happen to know the purpose of that function?  Why does it get called before I try to jump to the next/previous error?

**Comment by fice-t** [comment]
Ha, I actually forgot that bit. My bad. No worries though -- I didn't spend a lot of time here.

I don't think it's reasonably possible that haskell-mode can work around this. Also, it seems like `folding-mode` is dead, so I don't think anything will get done there any time soon.

I'm not sure of the specifics but it looks like the next/previous error function uses the overlays that the function in question creates.

@gracjan is there an easy way to disable this?

Related: #1170

**Comment by fice-t** [comment]
Not sure if it'll solve all problems, but you could do something like so:
```elisp
(defun haskell-check-paint-overlay (&rest args) nil)
```

**Comment by shanemikel** [comment]
As for working around, I was thinking there could be an option to disable, and some kind of state + hook pair that could call a registered function when the last compile encountered errors, and another when a compile following the errors encountered none.

The registered function would be whatever function the folding library exposes that expands all folds, and another to collapse them.  Whatever state is necessary to collapse only the previously collapsed folds is up to the folding library.

I know it might sound like a bit of an endeavor, but I'm assuming that the other folding plugins have the same problem.  If somebody can confirm this, I think it would be a worthy feature, and not very much trouble.

**Comment by shanemikel** [comment]
Just a simple dirty flag that gets reset on a clean compile

**Comment by shanemikel** [comment]
Okay, so I've evaluated `(defun haskell-check-paint-overlay (&rest args) nil)`, and haven't encountered any problems yet.

Now I understand the paint overlay thing.  I guess it's a little more complicated than notifying the folding library, the painting overlay is responsible for painting the error squigglies under the error points.  

So if there is to be a solution, it has to involve either haskell-mode notifying an intermediate library where to paint them, or haskell mode asking the intermediate library where is okay to paint and waiting for an event (buffer updated.. or folds changed) to ask/redraw again. (On top of the stuff I mentioned before)

**Comment by gracjan** [comment]
Hmm, I think that for the time being we should just disable the overlays. We rushed too fast with those and now there are issues we do not have resources to fix.

**Comment by gracjan** [comment]
Seems related: #1225.

**Comment by shanemikel** [comment]
Have we decided to have an option to disable these overlays yet?
